apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/commons/AbstractSamhsaMatcher.java

```diff
@@ -0,0 +1,378 @@
+package gov.cms.bfd.server.war.commons;
+
+import com.google.common.annotations.VisibleForTesting;
+import gov.cms.bfd.model.codebook.data.CcwCodebookVariable;
+import gov.cms.bfd.server.war.adapters.CodeableConcept;
+import gov.cms.bfd.server.war.adapters.Coding;
+import gov.cms.bfd.server.war.adapters.DiagnosisComponent;
+import gov.cms.bfd.server.war.adapters.ItemComponent;
+import gov.cms.bfd.server.war.adapters.ProcedureComponent;
+import gov.cms.bfd.sharedutils.exceptions.BadCodeMonkeyException;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.InputStreamReader;
+import java.io.UncheckedIOException;
+import java.nio.charset.StandardCharsets;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Set;
+import java.util.function.Predicate;
+import java.util.stream.Collectors;
+import org.apache.commons.csv.CSVFormat;
+import org.apache.commons.csv.CSVParser;
+import org.hl7.fhir.exceptions.FHIRException;
+
+public abstract class AbstractSamhsaMatcher<T> implements Predicate<T> {
+
+  /** The {@link CSVFormat} used to parse the SAMHSA-related code CSV files. */
+  private static final CSVFormat CSV_FORMAT = CSVFormat.EXCEL.withHeader();
+
+  protected static final String DRG =
+      CCWUtils.calculateVariableReferenceUrl(CcwCodebookVariable.CLM_DRG_CD);
+
+  private final List<String> drgCodes;
+  private final List<String> cptCodes;
+  private final List<String> icd9ProcedureCodes;
+  private final List<String> icd9DiagnosisCodes;
+  private final List<String> icd10ProcedureCodes;
+  private final List<String> icd10DiagnosisCodes;
```

Comment:
At least some of these would be better as Sets rather than Lists.  Also if they are immutable after loading it might be better to preload or lazy load them into a singleton and keep them around in memory rather than reinitialize them

Reply:
Yes, sets would bet more optimal, that was sort of a lazy copy over, I'll make that change